Bugs 1439142, 1417641 - network diagnostics fixes

pravisankar opened the pull request:
Bug 1439142 - Use openshift/origin-deployer image instead of openshift/hello-openshift as network diagnostic test pod.

Bug 1417641 - Make network diagnostic test pod image/protocol/port configurable

pravisankar commented:
@openshift/networking @sosiouxme PTAL

knobunc commented:
[test] not that it should catch anything.

knobunc commented:
https://bugzilla.redhat.com/show_bug.cgi?id=1439142 and https://bugzilla.redhat.com/show_bug.cgi?id=1417641

dcbw commented:
LGTM, but the jenkins failures are real and looks like you need to update the bash command completions too.

pravisankar commented:
[test]

knobunc commented:
[merge]

sosiouxme commented:
@pravisankar I deployed a 2-node origin cluster and ran the check successfully. Then I tried running it specifying an OSE image:

`oc adm diagnostics NetworkCheck --network-test-pod-image=registry.access.redhat.com/openshift3/ose-deployer`

This gave me:

```
       [Note] Running diagnostic: CheckPodNetwork
              Description: Check pod to pod communication in the cluster. In case of ovs-subnet network plugin, all pods should be able to communicate with each other and in case of multitenant network plugin, pods in non-global projects should be isolated and pods in global projects should be able to access any pod in the cluster and vice versa.
              
       WARN:  [DPodNet1005 from diagnostic CheckPodNetwork@openshift/origin/pkg/diagnostics/networkpod/pod.go:125]
              Skipping pod connectivity test for global projects on the same node. Reason: Couldn't find 2 global pods.
              
       WARN:  [DPodNet1005 from diagnostic CheckPodNetwork@openshift/origin/pkg/diagnostics/networkpod/pod.go:125]
              Skipping pod connectivity test for global projects on different nodes. Reason: Couldn't find 2 global pods.
```

This isn't as informative as I'd like. Docker pulled the image, so I think that part worked fine, but what happened when it tried to run? Why didn't the test pods run? Did I specify something wrong or is this a system problem?

pravisankar commented:
@sosiouxme 
`oc adm diagnostics NetworkCheck --network-test-pod-image=registry.access.redhat.com/openshift3/ose-deployer` failed because it is not listening on default test pod port 8080. I agree that there is no useful warning or info displayed on the terminal.

pravisankar commented:
@knobunc @dcbw @sosiouxme 

Updated, added couple of changes:
- On OCP, defaultImagePrefix will be 'openshift3/ose-' and it will not match full path 'registry.access.redhat.com/openshift3/ose-'. Now, we sanitize the test pod image before comparing with default diagnostics test pod image. 
- When network diagnostic test pods fails, you will notice an error or warning.

pravisankar commented:
flake https://github.com/openshift/origin/issues/14122
re[test]

knobunc commented:
[merge]

knobunc commented:
[merge][severity:blocker]

sosiouxme commented on pkg/diagnostics/network/setup.go:
```diff
@@ -152,23 +154,23 @@ func (d *NetworkDiagnostic) waitForTestPodAndService(nsList []string) error {
+				d.res.Warn("DNet3002", nil, fmt.Sprintf("Failed to run some network diags test pods: %d, So some network diagnostic checks may be skipped.", (totalPods-runningPods)))
 				return nil
+			} else {
+				errList = append(errList, fmt.Errorf("Failed to run network diags test pods, failed: %d, total: %d", (totalPods-runningPods), totalPods))
```
It would be really nice here if it could report on what happened with the ones that aren't running.

pravisankar replied:
Yes, We could pull the container logs in case of failures.

sosiouxme commented on pkg/diagnostics/network/setup.go:
```diff
@@ -177,15 +179,12 @@ func (d *NetworkDiagnostic) getNumOfValidTestPods(nsList []string, validPhases [
-					break
-				}
+			if pod.Status.Phase == kapi.PodRunning {
+				runningPodCount += 1
```
`} else { // build a list of pod state and any errors that can be retrieved for those that failed`

pravisankar commented on pkg/diagnostics/network/objects.go:
```diff
@@ -114,9 +112,26 @@ func GetTestPod(podName, nodeName string) *kapi.Pod {
+	} else {
+		trimmedPodImage = imageTokens[n-2] + "/" + imageTokens[n-1]
+	}
+	if trimmedPodImage == util.NetworkDiagDefaultTestPodImage {
```
Requirement for our network diagnostic test pod is to serve/listen for requests using tcp/udp on some port. We don't care if the test pod is running `socat` or simple http web server. We chose `socat` because `openshift/origin-deployer` doesn't serve by default and we picked this image for test pod because it already exists in redhat registry and also available on all the nodes. 
For non default test pod passed by user we expect it be serving by default.
